Source/WebCore/bindings/js/JSElementCustom.cpp

```diff
@@ -81,4 +85,62 @@ JSValue toJSNewlyCreated(JSGlobalObject*, JSDOMGlobalObject* globalObject, Ref<E
     return createNewElementWrapper(globalObject, WTFMove(element));
 }
 
+static JSValue getElementsArrayAttribute(JSGlobalObject& lexicalGlobalObject, const JSElement& thisObject, const QualifiedName& attributeName)
+{
+    auto& vm = JSC::getVM(&lexicalGlobalObject);
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    JSObject* cachedObject = nullptr;
+    JSValue cachedObjectValue = thisObject.getDirect(vm, builtinNames(vm).cachedAttrAssociatedElementsPrivateName());
+    if (cachedObjectValue)
+        cachedObject = asObject(cachedObjectValue);
+    else {
+        cachedObject = constructEmptyObject(vm, thisObject.globalObject()->nullPrototypeObjectStructure());
+        const_cast<JSElement&>(thisObject).putDirect(vm, builtinNames(vm).cachedAttrAssociatedElementsPrivateName(), cachedObject);
+    }
+
+    std::optional<Vector<RefPtr<Element>>> elements = thisObject.wrapped().getElementsArrayAttribute(attributeName);
```

Comment:
It seems wasteful to always call this function. Why not check the cache first?

Reply:
But what happens when this function starts returning a different value??

Reply:
We're checking that the cache is valid a few lines below, and only return the cached one if it's valid:
```cpp
        std::optional<Vector<RefPtr<Element>>> cachedElements = convert<IDLNullable<IDLFrozenArray<IDLInterface<Element>>>>(lexicalGlobalObject, cachedValue);
        if (elements == cachedElements)
            return cachedValue;
```

This implementation is following the proposed spec text (see https://whatpr.org/html/3917/common-dom-interfaces.html):
> * On getting, the IDL attribute must perform the following steps:
> 
>   1. Let elements be this element's [attr-associated elements](https://whatpr.org/html/3917/common-dom-interfaces.html#attr-associated-elements).
> 
>   2. If the contents of elements is equal to the contents of this element's [cached attr-associated elements](https://whatpr.org/html/3917/common-dom-interfaces.html#cached-attr-associated-elements), then return this element's [cached attr-associated elements](https://whatpr.org/html/3917/common-dom-interfaces.html#cached-attr-associated-elements).
> 
>   3. Let elementsAsFrozenArray be elements, [converted](https://heycam.github.io/webidl/#es-type-mapping) to a FrozenArray<T>?.
> 
>   4. Set this element's [cached attr-associated elements](https://whatpr.org/html/3917/common-dom-interfaces.html#cached-attr-associated-elements) to elementsAsFrozenArray.
> 
>   5. Return elementsAsFrozenArray.

Reply:
Oh, I missed that. Hm... but can this conversion have any script observable side effects? It looks like FrozenArray is just an array with `Object.freeze` so maybe not? length isn't configurable and none of the elements could be reconfigured either. It does seem a bit wasteful to convert the whole thing back to C++ just to check the inequality though... I don't think there is a way around it unfortunately. We'd either pay the memory cost or this runtime cost short of implementing FrozenArray semantics in C++.

Reply:
I didn't find a way to check the inequality of `JSValue` directly. If there's a way to compare `cachedValue === elementsValue`, we won't need to convert things to `std::optinal<Vectror>`; maybe I missed something and we have some `JSValue` method that could check that.

Reply:
We don't actually need to convert the FrozenArray to JS right?

One thing we could do is have a special JSCell type for the cache object which holds the FrozenArrays directly, and then we can avoid creating JS values until they're needed.

Just an idea